apps/managedidentity/managedidentity.go

```diff
@@ -135,65 +194,171 @@ func New(id ID, options ...ClientOption) (Client, error) {
 	default:
 		return Client{}, fmt.Errorf("unsupported type %T", id)
 	}
+
 	client := Client{
 		miType:     id,
 		httpClient: opts.httpClient,
+		source:     source,
 	}
+
+	client.authParams, err = createFakeAuthParams(client)
+	if err != nil {
+		return Client{}, err
+	}
+
 	return client, nil
 }
 
-func createIMDSAuthRequest(ctx context.Context, id ID, resource string, claims string) (*http.Request, error) {
-	var msiEndpoint *url.URL
-	msiEndpoint, err := url.Parse(imdsEndpoint)
-	if err != nil {
-		return nil, fmt.Errorf("couldn't parse %q: %s", imdsEndpoint, err)
+// GetSource detects and returns the managed identity source available on the environment.
+func getSource() (Source, error) {
+	identityEndpoint := os.Getenv(identityEndpointEnvVar)
+	identityHeader := os.Getenv(identityHeaderEnvVar)
+	identityServerThumbprint := os.Getenv(identityServerThumbprintEnvVar)
+	msiEndpoint := os.Getenv(msiEndpointEnvVar)
+	imdsEndpoint := os.Getenv(imdsEndVar)
+
+	if identityEndpoint != "" && identityHeader != "" {
+		if identityServerThumbprint != "" {
+			return ServiceFabric, nil
+		}
+		return AppService, nil
+	} else if msiEndpoint != "" {
+		return CloudShell, nil
+	} else if isAzureArcEnvironment(identityEndpoint, imdsEndpoint, runtime.GOOS) {
+		return AzureArc, nil
 	}
-	msiParameters := msiEndpoint.Query()
-	msiParameters.Set(apiVersionQuerryParameterName, imdsAPIVersion)
-	resource = strings.TrimSuffix(resource, "/.default")
-	msiParameters.Set(resourceQuerryParameterName, resource)
 
-	if len(claims) > 0 {
-		msiParameters.Set("claims", claims)
+	return DefaultToIMDS, nil
+}
+
+// Acquires tokens from the configured managed identity on an azure resource.
+//
+// Resource: scopes application is requesting access to
+// Options: [WithClaims]
+func (client Client) AcquireToken(ctx context.Context, resource string, options ...AcquireTokenOption) (base.AuthResult, error) {
+	o := AcquireTokenOptions{}
+	for _, option := range options {
+		option(&o)
 	}
 
-	switch t := id.(type) {
-	case UserAssignedClientID:
-		msiParameters.Set(miQueryParameterClientId, string(t))
-	case UserAssignedResourceID:
-		msiParameters.Set(miQueryParameterResourceId, string(t))
-	case UserAssignedObjectID:
-		msiParameters.Set(miQueryParameterObjectId, string(t))
-	case systemAssignedValue: // not adding anything
+	// ignore cached access tokens when given claims
+	if o.claims == "" {
+		if cacheManager == nil {
+			return base.AuthResult{}, errors.New("cache instance is nil")
```

Comment:
This error can never occur? App developer doesn't configure the cache for MSI. Why are we throwing?

Reply:
removed